crud/common/sharding/storage_metadata_cache.lua

```diff
@@ -0,0 +1,89 @@
+local stash = require('crud.common.stash')
+local utils = require('crud.common.sharding.utils')
+
+local storage_metadata_cache = {}
+
+local FUNC = 1
+local KEY = 2
+
+local cache_data = {
+    [FUNC] = nil,
+    [KEY] = nil,
+}
+
+local ddl_space = {
+    [FUNC] = '_ddl_sharding_func',
+    [KEY] = '_ddl_sharding_key',
+}
+
+local trigger_stash = stash.get(stash.name.ddl_triggers)
+
+local function update_sharding_func_hash(tuple)
+    local space_name = tuple[utils.SPACE_NAME_FIELDNO]
+    local sharding_func_def = utils.extract_sharding_func_def(tuple)
+    cache_data[FUNC][space_name] = utils.compute_hash(sharding_func_def)
+end
+
+local function update_sharding_key_hash(tuple)
+    local space_name = tuple[utils.SPACE_NAME_FIELDNO]
+    local sharding_key_def = tuple[utils.SPACE_SHARDING_KEY_FIELDNO]
+    cache_data[KEY][space_name] = utils.compute_hash(sharding_key_def)
+end
+
+local update_hash = {
+    [FUNC] = update_sharding_func_hash,
+    [KEY] = update_sharding_key_hash,
+}
+
+local function init_cache(section)
+    cache_data[section] = {}
+
+    local space = box.space[ddl_space[section]]
+
+    local update_hash_func = update_hash[section]
+
+    -- Remove old trigger if there was some code reload.
+    -- It is possible that ddl space was dropped and created again,
+    -- so removing non-existing trigger will cause fail;
+    -- thus we use pcall.
+    pcall(space.on_replace, space, nil, trigger_stash[section])
+
+    trigger_stash[section] = space:on_replace(
+        function(_, new)
+            return update_hash_func(new)
+        end
+    )
+
+    for _, tuple in space:pairs() do
+        local space_name = tuple[utils.SPACE_NAME_FIELDNO]
+        if cache_data[section][space_name] == nil then
+            update_hash_func(tuple)
+        end
+    end
```

Comment:
Duplicate entries? As I [see](https://github.com/tarantool/ddl/blob/f63b291738efeff3262432aab7948075a88cced2/ddl.lua#L95-L111) in tarantool/ddl the `_ddl_sharding_key` and `_ddl_sharding_func` spaces have primary (and so unique) indexes on the space name column.

Reply:
To be honest, I'm not sure how exactly `pairs` work and how it behaves if data is updated in the middle of the cycle with trigger.

If `cache_data[section][space_name]` is not nil, it is already filled with trigger. If we _may get_ outdated info from `pairs` it would lead to inconsistency. On the other hand, this approach ensures that we always have up-to-date info disregarding how pairs works.

Reply:
AFAIU we can't see a tuple with the same primary key during iteration by the primary index, at least since https://github.com/tarantool/tarantool/issues/1796.

Anyway, we don't do any DML here (trigger is a runtime object), spaces are in memtx, so we'll not yeilds during the iteration.

If we would do some DDL, we should run the loop in a transaction (if a transaction is not opened by a calling code). Here we don't need it, I guess.

I'm okay with any approach here, but, please, add your assumptions into a comment in the code.

Reply:
Added a comment